### PR TITLE
Enable folding beta 0 in XSMM flags

### PR DESCRIPTION
**Pull request by chelini**
This PR allows us to generate the same micro-kernel as DNN, which has
been checked manually for the `prepacked_targets.` However, there is
still some performance gap with DNN. Some observations in regard:
- There are no allocations in DNN. Everything is pre-allocated. We do
  have allocations in the kernels. For example, `gemm_fp32_mlir` ends up
  with two mallocs.
- Allocations have different alignments (64 for us while 2097152 for
  DNN).

### CODE REVIEW

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 1]
```diff
@@ -808,6 +814,100 @@ void ConvertLinalgToXsmm::runOnOperation() {
...
+                                                xsmm::GemmFlags::BETA_0));
+    gemmDispatchOp.setFlagsAttr(rewriter.getArrayAttr(newFlags));
+  });
+}
+
+static void fuseZeroWithGemmOrBrgemm(RewriterBase &rewriter,
+                                     xsmm::UnaryOp rootOp) {
+  LLVM_DEBUG(llvm::dbgs() << "[fuseZeroWithGemmOrBrgemm] Candidate op: "
+                          << rootOp << "\n");
+  // 1. Prove the alloc to be zero. Linear walk of the bb
```
Comment (adam-smnk) [inline lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:848]: There's no check that the same buffer is zeroed out. For example:
```
func.func @zero_test(%arg0: memref<1x32x32xf32>, %arg1: memref<1x32x32xf32>, %arg2: memref<32x32xf32>) {
  %c1_i64 = arith.constant 1 : i64
  %cst = arith.constant 0.000000e+00 : f32
  %alloc = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
  %0 = xsmm.unary.dispatch zero [32, 512, 1, 512] flags = (bcast_scalar) data_type = f32
  xsmm.unary zero(data_type = f32, %0, %cst, %alloc) : (i64, f32, memref<32x32xf32>) -> ()
  %1 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 32, 1] flags = (none) data_type = f32
  // Sets incorrectly flags = (beta_0) when zero acts on %alloc not %arg2
  xsmm.brgemm(data_type = f32, %1, %arg0, %arg1, %arg2, %c1_i64) : (i64, memref<1x32x32xf32>, memref<1x32x32xf32>, memref<32x32xf32>, i64) -> ()
  return
}
```

Not saying we need to handle weird patterns but just avoid wrong codegen.

**Review by adam-smnk — COMMENTED** [review]
Approach looks valid just needs to be more conservative.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 2]
```diff
@@ -866,14 +869,23 @@ static void fuseZeroWithGemmOrBrgemm(RewriterBase &rewriter,
   if (it == itEnd)
     return;
   while (++it != itEnd) {
-    if (mlir::isPure(&*it))
+    // Skip pure operations and operations that do not
+    // touch `dest`.
```
Comment (adam-smnk) [inline lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:873]: Not sure if that's safe in case of aliasing.

Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:873]: if we have a subview on `dest` we fails.

Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:873]: Do you have in mind a specific pattern?

Comment (adam-smnk) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:873]: > if we have a subview on dest we fails.

Why would we, it is pure afaik and touches dest.

For example:
```
func.func @zero_test(%arg0: memref<1x32x32xf32>, %arg1: memref<1x32x32xf32>, %arg2: memref<32x32xf32>) {
  %c1_i64 = arith.constant 1 : i64
  %cst = arith.constant 0.000000e+00 : f32
  %alloc = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
  %0 = xsmm.unary.dispatch zero [32, 512, 1, 512] flags = (bcast_scalar) data_type = f32
  xsmm.unary zero(data_type = f32, %0, %cst, %alloc) : (i64, f32, memref<32x32xf32>) -> ()
  %sub = memref.subview %alloc[0, 0] [16, 16] [1, 1] : memref<32x32xf32> to memref<16x16xf32, strided<[32, 1]>>
  call @test(%sub) : (memref<16x16xf32, strided<[32, 1]>>) -> ()
  // Inserts beta_0 while we cannot prove it is zero anymore.
  %1 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 32, 1] flags = (none) data_type = f32
  xsmm.brgemm(data_type = f32, %1, %arg0, %arg1, %alloc, %c1_i64) : (i64, memref<1x32x32xf32>, memref<1x32x32xf32>, memref<32x32xf32>, i64) -> ()
  return
}
func.func private @test(%arg0 : memref<16x16xf32, strided<[32, 1]>>)
```

Comment (adam-smnk) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]: This still is fails if `memref.subview` is created before xsmm zero op.

Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]: Not sure if I get it. You can zero a subview, as long as you don't do something with it before the GEMM.

File: test/Integration/xsmm-strided-brgemm3.mlir [hunk 3]
```diff
@@ -24,12 +25,12 @@ func.func @matmul_static(%A : !A_tensor_t, %B : !B_tensor_t, %C : !C_tensor_t) {
...
   %fill = linalg.fill ins(%cst_fill : f32) outs(%empty: tensor<2x2x8x2xf32>) -> tensor<2x2x8x2xf32>
 
-  // IR: %[[C2:.+]] = arith.constant 2 : i64
+  // IR-DAG: %[[C2:.+]] = arith.constant 2 : i64
   // IR-DAG: %[[C1:.+]] = arith.constant 1 : i64
   // IR-DAG: %[[C8:.+]] = arith.constant 8 : i64
   // IR-DAG: %[[C4:.+]] = arith.constant 4 : i64
   // IR-DAG: %[[C0:.+]] = arith.constant 0 : i64
-  // IR: xsmm_gemm_dispatch(%[[C1]], %[[C8]], %[[C2]], %[[C4]], %[[C8]], %[[C2]], %[[C2]], %[[C0]]) 
+  // IR: xsmm_gemm_dispatch(%[[C1]], %[[C8]], %[[C2]], %[[C4]], %[[C8]], %[[C2]], %[[C2]], %[[C4]])
```
Comment (rengolin) [inline test/Integration/xsmm-strided-brgemm3.mlir:33]: I'd also add here a check for NOT zero

File: test/Passes/DefaultPipeline/linalg-to-xsmm.mlir [hunk 4]
```diff
@@ -30,3 +30,34 @@ func.func @fill_op_i32(%arg0: memref<3x3xi32>) {
...
+  return %mul : tensor<3x3xf32>
+}
+
+// CHECK-LABEL: gemm_with_zero
+// CHECK-SAME: %[[ARG0:.+]]: memref<3x3xf32>, %[[ARG1:.+]]: memref<3x3xf32>
+// CHECK-DAG: %[[C0:.+]] = arith.constant 0 : index
+// CHECK-DAG: %[[C1:.+]] = arith.constant 1 : i64
+// CHECK-DAG: %[[C3:.+]] = arith.constant 3 : i64
+// CHECK-DAG: %[[C4:.+]] = arith.constant 4 : i64
+// CHECK: %[[ALLOC:.+]] = memref.alloc() {alignment = 64 : i64} : memref<3x3xf32>
```
Comment (rengolin) [inline test/Passes/DefaultPipeline/linalg-to-xsmm.mlir:52]: // CHECK-NOT: zero

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp (hunk 2 above)
Comment (adam-smnk) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]:
```
  %alloc = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
  %sub = memref.subview %alloc[0, 0] [16, 16] [1, 1] : memref<32x32xf32> to memref<16x16xf32, strided<[32, 1]>>
  %0 = xsmm.unary.dispatch zero [32, 512, 1, 512] flags = (bcast_scalar) data_type = f32
  xsmm.unary zero(data_type = f32, %0, %cst, %alloc) : (i64, f32, memref<32x32xf32>) -> ()
  call @test(%sub) : (memref<16x16xf32, strided<[32, 1]>>) -> ()
  %1 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 32, 1] flags = (none) data_type = f32
```

In this pattern, folding shouldn't kick in as `%alloc` is modified through an alias.

**Comment by rengolin** [comment]
I don't see a big difference in the benchmarks, are you sure this pass is triggered on the benchmarks?

**Comment by chelini** [comment]
> > I don't see a big difference in the benchmarks, are you sure this pass is triggered on the benchmarks?
> 
> Yes, it does. I checked the generated LIBXSMM kernel.

at least in the ` prepacked_targets`.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 5]
```diff
@@ -808,6 +814,120 @@ void ConvertLinalgToXsmm::runOnOperation() {
     return signalPassFailure();
 }
 
+template <typename OpTy>
+static void updateGemmOpFlags(RewriterBase &rewriter, OpTy gemmDispatchOp) {
+  static_assert(llvm::is_one_of<OpTy, xsmm::GemmDispatchOp,
+                                xsmm::BrgemmDispatchOp>::value);
+  if (!gemmDispatchOp->hasOneUse())
```
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:821]: There is one right after LICM. If we have layers of the same size, it does remove the later `dispatch` calls.

Reproduce with:
```
mlir-gen --layers=1024,1024,1024,1024 | tpp-opt --default-tpp-passes
```

and see the third `scf.parallel` reusing the dispatch from the previous layer (`%5` and `%6`).

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 6]
```diff
@@ -808,6 +814,120 @@ void ConvertLinalgToXsmm::runOnOperation() {
...
+                          << rootOp << "\n");
+  // 1. Prove the alloc to be zero. Linear walk of the bb
+  // and make sure there are only pure operations between
+  // the zero op and the gemm.
+  Value dest = rootOp.getInputs().back();
+  DenseSet<Operation *> destUsers(dest.getUsers().begin(),
+                                  dest.getUsers().end());
+
+  Block *blck = nullptr;
+  if (auto bbArg = dyn_cast<BlockArgument>(dest)) {
```
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:857]: So, the function is called `fuseZeroWithGemmOrBrgemm`, and the walk is about making sure they're side-effect free zeroes, not fusing. The actual logic is below (line 913). There's a lot to unpack in the walk, what each match means, why some continue others return, etc.

I'd try to find common patterns, create list holders for the matches and return if none were found, etc.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:821]: > If we allow CSE we need to start cloning the operations.

No, just run before CSE or LICM would do. Then different `dispatch`es won't get commoned up.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 7]
```diff
@@ -808,6 +814,129 @@ void ConvertLinalgToXsmm::runOnOperation() {
...
+void FoldXsmmFlags::runOnOperation() {
+  SmallVector<xsmm::UnaryOp> producers;
+  getOperation()->walk([&](xsmm::UnaryOp unaryOp) {
+    auto kind = unaryOp.getCallee();
+    if (kind == xsmm::UnaryKind::ZERO)
+      producers.push_back(unaryOp);
+  });
+
+  IRRewriter rewriter(&getContext());
+  for (xsmm::UnaryOp producer : producers)
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]: Done.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp [hunk 8]
```diff
@@ -808,6 +814,120 @@ void ConvertLinalgToXsmm::runOnOperation() {
...
+  }
+  if (it == itEnd)
+    return;
+
+  LLVM_DEBUG(llvm::dbgs() << "[fuseZeroWithGemmOrBrgemm] Candidate op OK: "
+                          << rootOp << "\n");
+
+  // 2. Update flags.
+  assert(isa<xsmm::GemmOp>(*it) || isa<xsmm::BrgemmOp>(*it));
+  if (auto gemmOp = dyn_cast<xsmm::GemmOp>(*it)) {
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]: Added case for fusedBrgemm.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp:865]: I tried to pull it out. Let me know what you think.

File: lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp]: thanks, fixed.

File: include/TPP/Dialect/Xsmm/XsmmOps.td [hunk 9]
```diff
@@ -81,6 +81,10 @@ def Xsmm_GemmOp : Xsmm_Op<"gemm"> {
     `(` `data_type` `=` $data_type `,` $inputs `)`
     attr-dict `:` functional-type($inputs, results)
   }];
+
+  let extraClassDeclaration = [{
+    Value getOutput() { return getInputs().back(); }
```
Comment (rengolin) [inline include/TPP/Dialect/Xsmm/XsmmOps.td:86]: Nit: to be in sync with the others, maybe use `getInputs()[3]` too?

**Review by rengolin — APPROVED** [review]
Small nit, but LGTM, thanks!

**Comment by chelini** [comment]
@adam-smnk any more comments?

**Review by adam-smnk — APPROVED** [review]
Looks good, thanks 👍

**Pull request merged**